src/signalrclient/signalr_default_scheduler.cpp

```diff
@@ -0,0 +1,217 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "stdafx.h"
+#include <assert.h>
+#include "signalr_default_scheduler.h"
+#include <thread>
+
+namespace signalr
+{
+    thread::thread()
+        : m_internals(std::make_shared<internals>())
+    {
+        auto internals = m_internals;
+
+        m_thread = std::thread([=]()
+            {
+                while (true)
+                {
+                    {
+                        signalr_base_cb cb;
+                        {
+                            std::unique_lock<std::mutex> lock(internals->m_callback_lock);
+                            auto& callback = internals->m_callback;
+                            auto& closed = internals->m_closed;
+
+                            if (closed && callback == nullptr)
+                            {
+                                assert(callback == nullptr);
+                                return;
+                            }
+
+                            if (callback == nullptr)
+                            {
+                                internals->m_callback_cv.wait(lock,
+                                    [&callback, &closed]
+                                    {
+                                        return closed || callback != nullptr;
+                                    });
+                            }
+
+                            if (closed && callback == nullptr)
+                            {
+                                assert(callback == nullptr);
+                                return;
+                            }
+
+                            cb = callback;
+                            internals->m_callback = nullptr;
+                        } // unlock
+
+                        try
+                        {
+                            cb();
+                        }
+                        catch (...)
+                        {
+                            // ignore exceptions?
+                        }
+                    } // destruct cb, otherwise it's possible a shared_ptr is being held by the lambda/function and on destruction it could schedule work which could be
+                    // added to this thread and if it then blocks waiting for completion it would deadlock. If we destruct before setting the m_busy flag,
+                    // another thread will run the work instead.
+
+                    internals->m_busy = false;
+                }
+
+                assert(internals->m_callback == nullptr);
+            });
+    }
+
+    void thread::add(signalr_base_cb cb)
+    {
+        {
+            assert(m_internals->m_closed == false);
```

Comment:
Any risk for not checking these asserts while not holding the lock?

Reply:
These specific ones shouldn't have any risk as the closed flag is set in the destructor and the busy flag is checked before calling `add`.